src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,416 @@
+
+                // TODO: https://github.com/dotnet/roslyn/issues/42235
+                // The type of lambda function is null, but it probably can be matched to its corresponding parameter 
+                // Keep an eye on the identifier that is syntax and semantic true but the type is null.
+                if (argType.Type == null)
+                    return false;
+
+                // rule (1)
+                if (!inOrder && nameSyntax is null)
+                    return false;
+
+                // The type of argument must be convertible to the type of parameter
+                if (!parameters[parameterIndex].IsParams
+                    && semanticModel.Compilation.ClassifyCommonConversion(argType.Type, parameters[parameterIndex].Type).Exists)
+                {
+                    // rule (3)
+                    if (matchedTypes[parameterIndex])
+                        return false;
+                    matchedTypes[parameterIndex] = true;
+                }
+                else if (parameters[parameterIndex].IsParams
+                    && semanticModel.Compilation.ClassifyCommonConversion(argType.Type, parameters[parameterIndex].Type).Exists)
+                {
+                    // The parameter with keyword params takes an array type, then it cannot be matched more than once
+                    // rule (3)
+                    if (matchedTypes[parameterIndex])
+                        return false;
+                    matchedTypes[parameterIndex] = true;
+                    paramsMatchedByArray = true;
+                }
+                else if (parameters[parameterIndex].IsParams
+                         && parameters.Last().Type is IArrayTypeSymbol paramsType
+                         && semanticModel.Compilation.ClassifyCommonConversion(argType.Type, paramsType.ElementType).Exists)
+                {
+                    // The parameter with keyword params can take a *variable* number of arguments,
+                    // compare its element type with the argument's type.
+
+                    // rule (3)
+                    if (matchedTypes[parameterIndex] && paramsMatchedByArray)
+                        return false;
+                    matchedTypes[parameterIndex] = true;
+                    paramsMatchedByArray = false;
+                }
+                else
+                {
+                    // rule (2)
+                    return false;
+                }
+
+                if (targetArgument.Equals(arguments[i]))
+                    targetParamIndex = parameterIndex;
+            }
+
+            // mark all optional parameters as matched
+            // rule (4)
+            for (var i = 0; i < parameters.Length; i++)
+            {
+                if (parameters[i].IsOptional || parameters[i].IsParams)
+                {
+                    matchedTypes[i] = true;
+                }
+            }
+
+            // rule (4)
+            return matchedTypes.All(a => a);
+        }
```

Comment:
so this method is enormous an scary.  Have we considered something *much* simpler.  namely, if you have a method call to something like `Foo(a, b, c)`, we simply attempt the rewrite to `Foo(a, (D)b, c)` and see if that binds properly now?  This is possible with speculative semantic models.  I can walk youthrough how to do this.  It seems like it would just avoid a ton of having to reimplement this compiler logic.

Reply:
Yup, I replace old function by this in my up-to-date commit :)

Reply:
woohoo!